flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala

```diff
@@ -18,18 +18,26 @@
 package org.apache.flink.table.planner.codegen
 
 import org.apache.flink.configuration.Configuration
-import org.apache.flink.table.data.GenericRowData
-import org.apache.flink.table.types.logical.{BigIntType, IntType, RowType, VarBinaryType}
+import org.apache.flink.table.data.{GenericArrayData, GenericMapData, GenericRowData, StringData}
+import org.apache.flink.table.types.logical.{ArrayType, BigIntType, IntType, MapType, MultisetType, RowType, VarBinaryType, VarCharType}
 
-import org.junit.{Assert, Test}
+import org.junit.{Assert, Rule, Test}
+import org.junit.rules.ExpectedException
+
+import scala.collection.JavaConversions.mapAsJavaMap
 
 /** Test for [[HashCodeGenerator]]. */
 class HashCodeGeneratorTest {
 
   private val classLoader = Thread.currentThread().getContextClassLoader
 
+  var expectedException: ExpectedException = ExpectedException.none
```

Comment:
Would you please add a ITCase in SQL to cover group by array type fields and map type fields?

Reply:
I have add an ITCase about array type in group by clause.

Reply:
I have researched spark and presto. In spark, map type is not allowed used in join and group by clause because of map type is not an orderable type, but it supported as the distribute by key. In presto, map supported as group by key, it is because presto only implement hash-based aggregate. So, I think we should support map type field as shuffle key, but doesn't as sort key, such as order by/sort-base join/agg clause. we should support to generate hash code for map which is also as spark and presto, this may be useful for distribute by clause which we will support in 1.16.